drivers/sensor/pixart/pat9136/pat9136_stream.c

```diff
@@ -0,0 +1,431 @@
+			   buf_len_required, err);
+
+	/** Still throw an error even if asserts are off */
+	if (err) {
+		struct rtio_iodev_sqe *iodev_sqe = data->stream.iodev_sqe;
+
+		LOG_ERR("Failed to acquire buffer for encoded data: %d", err);
+
+		data->stream.iodev_sqe = NULL;
+		rtio_iodev_sqe_err(iodev_sqe, err);
+		return;
+	}
+
+	struct rtio_sqe *write_res_x_sqe = rtio_sqe_acquire(data->rtio.ctx);
+	struct rtio_sqe *read_res_x_sqe = rtio_sqe_acquire(data->rtio.ctx);
+	struct rtio_sqe *write_res_y_sqe = rtio_sqe_acquire(data->rtio.ctx);
+	struct rtio_sqe *read_res_y_sqe = rtio_sqe_acquire(data->rtio.ctx);
+	struct rtio_sqe *write_sqe = rtio_sqe_acquire(data->rtio.ctx);
+	struct rtio_sqe *read_sqe = rtio_sqe_acquire(data->rtio.ctx);
+	struct rtio_sqe *cb_sqe = rtio_sqe_acquire(data->rtio.ctx);
+	uint8_t val;
+
+	err = sensor_clock_get_cycles(&cycles);
+	CHECKIF(err) {
+		struct rtio_iodev_sqe *iodev_sqe = data->stream.iodev_sqe;
+
+		LOG_ERR("Failed to get timestamp: %d", err);
+
+		data->stream.iodev_sqe = NULL;
+		rtio_iodev_sqe_err(iodev_sqe, err);
+		return;
+	}
+	buf->header.timestamp = sensor_clock_cycles_to_ns(cycles);
+
+	CHECKIF(!write_res_x_sqe || !read_res_x_sqe ||
+		!write_res_y_sqe || !read_res_y_sqe ||
+		!write_sqe || !read_sqe || !cb_sqe) {
+		LOG_ERR("Failed to acquire RTIO SQE's");
+
+		data->stream.iodev_sqe = NULL;
+		rtio_iodev_sqe_err(data->stream.iodev_sqe, -ENOMEM);
+		return;
+	}
+
+	/* X Resolution used for decoding DXY in mm */
+	{
+		val = REG_RESOLUTION_X_LOWER | REG_SPI_READ_BIT;
+
+		rtio_sqe_prep_tiny_write(write_res_x_sqe,
+					 data->rtio.iodev,
+					 RTIO_PRIO_HIGH,
+					 &val,
+					 1,
+					 NULL);
+		write_res_x_sqe->flags |= RTIO_SQE_TRANSACTION;
+
+		rtio_sqe_prep_read(read_res_x_sqe,
+				   data->rtio.iodev,
+				   RTIO_PRIO_HIGH,
+				   &buf->header.resolution.buf[0],
+				   1,
+				   NULL);
+		read_res_x_sqe->flags |= RTIO_SQE_CHAINED;
+	}
+
+	/* Y Resolution used for decoding DY in mm */
+	{
+		val = REG_RESOLUTION_Y_LOWER | REG_SPI_READ_BIT;
+
+		rtio_sqe_prep_tiny_write(write_res_y_sqe,
+					 data->rtio.iodev,
+					 RTIO_PRIO_HIGH,
+					 &val,
+					 1,
+					 NULL);
+		write_res_y_sqe->flags |= RTIO_SQE_TRANSACTION;
+
+		rtio_sqe_prep_read(read_res_y_sqe,
+				   data->rtio.iodev,
+				   RTIO_PRIO_HIGH,
+				   &buf->header.resolution.buf[2],
+				   1,
+				   NULL);
+		read_res_y_sqe->flags |= RTIO_SQE_CHAINED;
+	}
+
+	/* Pull out data */
+	{
+		val = REG_BURST_READ | REG_SPI_READ_BIT;
+
+		rtio_sqe_prep_tiny_write(write_sqe,
+					 data->rtio.iodev,
+					 RTIO_PRIO_HIGH,
+					 &val,
+					 1,
+					 NULL);
+		write_sqe->flags |= RTIO_SQE_TRANSACTION;
+
+		rtio_sqe_prep_read(read_sqe,
+				   data->rtio.iodev,
+				   RTIO_PRIO_HIGH,
+				   buf->buf,
+				   sizeof(buf->buf),
+				   NULL);
+		read_sqe->flags |= RTIO_SQE_CHAINED;
+	}
+
+	rtio_sqe_prep_callback_no_cqe(cb_sqe,
+				      pat9136_complete_result,
+				      (void *)dev,
+				      buf);
+
+	rtio_submit(data->rtio.ctx, 0);
```

Comment:
neat to see the sequence of reads here all chained up to a callback for the state machine, glad to see it is useful

Reply:
Yep - Moreover, the error handling stuff @FlorianWeber1018 is working on (#81014) will make it even better!